Don't request chunk unloads during fill

Puremin0rez opened the pull request:
Requesting chunk unloads during the fill task takes a tremendous hit on the server TPS, which leads to a really bad experience while players are online. It also slows down the fill process overall. By removing the chunk unload request, the server TPS will stay at a near 20 TPS due to the server now unloading the chunks naturally.

This leads to no noticeable downsides in my personal testing as well as many others personal testing over at the PaperMC discord. This will simply allow the server to handle unloading itself, which is seemingly much faster than a plugin requesting unloads.

I attempted to follow your comment style and commented out the line rather than deletion.

As always, feel free to do your own testing :)

Northbadge committed:
Improved fill progress saving & restoring

+ Added fill progress save to autosave, to avoid lost progress on hard-crashes
+ Prevented fill progress from being deleted except when properly finished or terminated.
+ In case fill progress is saved, but world is not loaded yet when config is loaded, running /wb reload should now resume filling if the world is loaded then.
+ Updated restoreprogress default values to match the current defaults.

Puremin0rez committed:
Don't request chunk unloads during fill
Requesting chunk unloads during the fill task takes a tremendous hit on the server TPS, which leads to a really bad experience while players are online. It also slows down the fill process overall. By removing the chunk unload request, the server TPS will stay at a near 20 TPS due to the server now unloading the chunks naturally.

clrxbl commented:
Definitely has improved the speed & TPS, but can easily run out of memory with this patch.

Puremin0rez commented:
> 
> 
> Definitely has improved the speed & TPS, but can easily run out of memory with this patch.

Just to verify - did you directly test against WorldBorder with and without this PR? It's easy to run out of memory with standard WorldBorder as well since I think there may be a memory leak somewhere (either within WB or MC itself - refer to the multiple open issues about memory).

While this change will lead to chunks being loaded slightly longer - the difference is like 1-2 seconds longer, and I don't really see how you could run out of memory since the chunks are still being unloaded - we're just not requesting their unload immediately.

Let me know what heap size you used and any other information that may be helpful so I can see if I can reproduce anything.

Brettflan commented:
> can easily run out of memory with this patch.

This is of course the reason WorldBorder requests the server to unload chunks that no longer need to be loaded after generation. Way back when this plugin was initially written, this was absolutely needed to keep from running out of memory due to the server just keeping the chunks loaded.
This was looked into again briefly a couple of years ago and found to still be just as necessary.

@Puremin0rez 
It's possible that Spigot/Paper somehow better decides to go ahead and unload chunks on its own during the process now, but I'd frankly be surprised if it kept up well enough for this change to be preferable for most servers. And if it really is working better in your experience for large Fill tasks, is that with Paper only, or Spigot as well? Also, how much memory do you have allocated to Paper/Spigot?

Puremin0rez commented:
I can completely understand why the force unloads were added - Minecraft used to leak chunks like crazy and forcing it to unload them was essentially required.

MC 1.14 changed the chunk system quite significantly and everything uses tickets now (as you are aware), and it appears that MC is good at unloading them all or at least doing the chunk cleanup by itself.

The actual problem this PR is trying to solve is that with the unload requests happening, it's forcing the chunk unloads sync (right now, right this moment) rather than vanilla doing it. This causes a significant TPS drop while a fill is running, essentially making filling on a live server not easily doable. I have no idea why the chunk unload request is so heavy - because you'd think it'd be fine, but it's clearly doing something that's quite intensive.

Majority of my testing was done on Spigot (as it's the upstream / most popular) but results on Paper were pretty much the same. All testing was done with 2GB (what I consider minimum), 5GB (average server?) and 10GB (larger server) heaps. I observed memory usage and chunks that remained loaded and while yes, chunks WILL remain loaded longer than if the chunks were not being forcibly unloaded, it was never more than 2 seconds at most in my testing in my personal environment. This might change based on storage speed, cpu speed, etc. It will also lead to an increase in memory as it's possible for chunks to remain loaded for slightly longer - but the memory increase wasn't even that significant even in my 2GB heap testing. I did not use memory dump analyzers or anything crazy like that.

The main takeaway for me was that there were never any leaked chunks / chunks that remained loaded even after a huge fill (I did 50k x 50k which took nearly a week).

I encourage you to to do your own testing when you get a chance and see what you think, as it's really good to have this tested in different environments. If you don't believe this to be preferred for all setups - it could always be thrown behind a config option for severe memory limited cases.

Thanks for reading!

bartyrealms commented:
Am doing testing on this currently, 10gb of RAM allocated.

It appears to be about the same speed as regular WorldBorder. However, there is no loss of speed present in this PR as there is in the regular WorldBorder plugin. I assume this is due to the unload function no longer being called (since that was the only change.

After about five minutes:
![image](https://user-images.githubusercontent.com/42015460/72683453-81ce0980-3aa5-11ea-8eb1-12d6a42b9b67.png)
Typically there is already a loss of speed noticed at this point.

TPS during the fill process:
![image](https://user-images.githubusercontent.com/42015460/72683466-927e7f80-3aa5-11ea-9f64-4f315085d7af.png)

Clearly this pull request solves the issue upon initial testing. I am currently generated a 10k x 10k world and will report back with my findings.

egg82 commented on src/main/java/com/wimbli/WorldBorder/WorldFillTask.java:
```diff
@@ -249,7 +249,8 @@ public void run()
 			if (!chunkOnUnloadPreventionList(unload.x, unload.z))
 			{
 				world.setChunkForceLoaded(unload.x, unload.z, false);
-				world.unloadChunkRequest(unload.x, unload.z);
```
Use `Config.AvailableMemoryTooLow()` to check if memory is low before calling this loop instead of removing this line.

egg82 reviewed (commented):
You can have the best of both worlds and give servers with larger memory capacity the benefit of having the server naturally unload chunks, and servers with less available memory the method they need to prevent overflows.

Wrap the chunk freeing loop in an if statement that can determine when to release memory. Along with the included memory checking method later in the file, this should prevent nasty crashes and still give a nice speed boost. The performance hit of the Runtime methods should still be far, far less than the performance hit of a chunk save (forcing disk I/O)

Take some inspiration from this method: https://github.com/Puremin0rez/WorldBorder/blob/4a16d5321a41c99858a4be30389d76a51b7b5875/src/main/java/com/wimbli/WorldBorder/Config.java#L536

TheCompieter commented:
i like this change it no longer indeed causes tps lagg on servers with more ram

dcoshea committed:
Correctly teleport to highest block instead of spawn (fixes #178)

The fix for issue #57 added some limits to the values searched when looking for
a safe Y coordinate for the teleport destination, including calling
getHighestBlockYAt() to determine the Y coordinate of the highest non-empty
block.  In non-nether cases with normal terrain that doesn't approach the
world's maximum Y values, highestBlockBoundary will be set to 1 greater than the
return value of getHighestBlockYAt(), i.e. the Y coordinate of the empty block
above the highest non-empty block.

In the non-nether, non-flying case, limTop was previously set to this value, and
the safe Y coordinate search would proceed up to *but not including* this limit,
so the maximum Y coordinate searched to check if it was empty would be the
highest *non*-empty y coordinate.  However, the player's current Y coordinate is
checked first even if it is equal to limTop (unless it is somehow not greater
than limBot).  As a result, if the only safe block for the player to be
teleported on top of was getHighestBlockYAt(), they would only be teleported
there if it wouldn't result in a change in their Y coordinate, otherwise they
would be teleported to spawn.

This fix expands the search to include highestBlockBoundary.

highestBlockBoundary cannot be greater than world.getMaxHeight() - 2, so the
search for a Y coordinate will now include that value rather than excluding it.
isSafeSpot() only adds 1 to that value, giving world.getMaxHeight() - 1, which
is still within the valid Y coordinate range.

This might also fix issue #115.

dcoshea committed:
Allow teleporting player partially or fully above the world

For worlds other than the nether (which is a special case due to the
bedrock ceiling) the maximum Y coordinate to which a player would be
teleported was previously world.getMaxHeight() - 2.  This enabled
isSafeSpot() to check the blocks at Y and Y + 1 to ensure that they
were breathable.  However, this meant that if there were blocks all
the way up to 1 or 2 less than world.getMaxHeight(), the player would
never be teleported on top of them, and instead would be teleported to
an air gap somewhere below them, or to spawn if no such gap was found.

This commit enables the player to be teleported to Y coordinates up to
world.getMaxHeight() so that they can be teleported on top of the
highest possible blocks (which would be at world.getMaxHeight() - 1).
As part of this, isSafeSpot() is updated to treat the void at Y >=
world.getMaxHeight() as breathable.

edwerd commented:
Does anyone have a build of this I can use? I'm having the same issue on my Minecraft server.

compieter-gh commented:
@edwerd 
https://github.com/TheCompieter/WorldBorder/releases/tag/v1.10.1

TheoRetisch1 commented:
@Puremin0rez will you update your awesome fork to version 1.16.1? :) He has been my faithful companion for the last months, thanks! Or does everybody use Chunkmaster these days?

Puremin0rez commented:
> 
> 
> @Puremin0rez will you update your awesome fork to version 1.16.1? :) He has been my faithful companion for the last months, thanks! Or does everybody use Chunkmaster these days?

No update is needed, everything remains completely functional on 1.16.

It seems many people use Chunkmaster, but I personally find it to be slower even when configured to be "fast"... so... I'll keep using the true and tested WorldBorder.

I will probably do an actual proper fork and maintain it going forward though (and provide downloads, etc) since I see quite a few people sharing "updated" WorldBorders that don't even have this PR in them, which is why people prefer ChunkMaster since it's quite a bit faster than the vanilla, standard WorldBorder.

docwhomc commented:
Time to get everyone with a WorldBorder fork together to unify and officially maintain the project suppose...

docwhomc commented:
@Puremin0rez do you have a discord?

docwhomc commented:
Well, if you do have one, I set up a WorldBorder Reboot Discord, which you can join [here](https://discord.gg/ePNBgMG).

Puremin0rez committed:
Merge pull request #1 from Brettflan/master

Upstream Merge

Puremin0rez committed:
Merge pull request #2 from dcoshea/y-coordinate-fixes

Height Teleport Fixes PR

Puremin0rez committed:
Merge pull request #3 from Northbadge/master

Improved fill progress saving & restoring PR

Puremin0rez commented:
I will be maintaining my own fork of WorldBorder going forward.

Thank you to the wonderful Brettflan for all of his years of hard work and dedication to the science behind world gen and creating a valuable tool for everyone.

**You may find it here:** https://github.com/Puremin0rez/WorldBorder
